plugins/bookmarks.js

```diff
@@ -734,6 +734,82 @@ window.plugin.bookmarks.loadStorageBox = function() {
     }
   }
 
+  window.plugin.bookmarks.optBookmarkVisible = function(command) {
+    // Being a bulk operation, this explicitly avoids hooks that happen on a
+    // per portal basis.  Instead, it uses the same one ran when bookmarks are
+    // imported.
+    console.log('BOOKMARKS: visible ' + command);
+    const displayBounds = map.getBounds();
+    const folders = window.plugin.bookmarks.bkmrksObj['portals'];
+    const idsInUse = new Set();
+
+    const counts = {'skip': 0, 'add': 0, 'delete': 0};
+    let total = 0;
+
+    // Find existing ids in use so we don't accidentally reuse them.
+    for (const idFolders of Object.keys(folders)) {
+      idsInUse.add(idFolders);
+      for (const idBkmrk of Object.keys(folders[idFolders]['bkmrk'])) {
+        idsInUse.add(idBkmrk);
+      }
+    }
+    for (const [guid, portal] of Object.entries(portals)) {
+      // The check for _map restricts to portals actually shown currently
+      if (displayBounds.contains(portal.getLatLng()) && portal._map) {
+        total += 1;
+
+        // First, figure out what to do with the portal.
+        let op = 'skip';
+        const bkmrkData = window.plugin.bookmarks.findByGuid(guid);
+        if (bkmrkData && ['off', 'toggle'].includes(command)) {
+          op = 'delete';
+        } else if (!bkmrkData && ['on', 'toggle'].includes(command)) {
+          op = 'add';
+        }
+        counts[op]++;
+
+        // Then do it.
+        switch (op) {
+          case 'skip':
+            break;
+
+          case 'add':
```

Comment:
When I first wrote this, there was a definite problem with collisions from `generateID()`.  Back then it was only `time+rand(0,99)`, and so bookmarking more than a handful of portals/second would cause problems.

It looks like that was fixed with https://github.com/iitc-project/ingress-intel-total-conversion/commit/3527cf1acb0e1e816ee49ac9eaa9d4cb948f5c50 .  I'd noticed in my testing I was not seeing any collisions logged.  Now I know why!

Addressing that should remove some of those nested loops.  They were just collecting existing IDs.

Reply:
Oh, as a bonus, removing the collision checking seems to have improved throughput from about 100 portals/sec to 500 portals/sec.

Reply:
Ugh.  Ignore all comments about speed.

Almost all of that time is spent in `jquerySortableScript()`.